src/java_bytecode/java_bytecode_instrument.cpp

```diff
@@ -0,0 +1,520 @@
+    if(code_ifthenelse.else_case().is_not_nil())
+      instrument_code(code_ifthenelse.else_case());
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_switch)
+  {
+    code_blockt block;
+    code_switcht &code_switch=to_code_switch(code);
+    block.copy_to_operands(
+      instrument_expr(code_switch.value()));
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_return)
+  {
+    if(code.operands().size()==1)
+    {
+      code_blockt block;
+      block.copy_to_operands(instrument_expr(code.op0()));
+      block.copy_to_operands(code);
+      code=block;
+    }
+  }
+  else if(statement==ID_function_call)
+  {
+    code_blockt block;
+    code_function_callt &code_function_call=to_code_function_call(code);
+    block.copy_to_operands(instrument_expr(code_function_call.lhs()));
+    block.copy_to_operands(instrument_expr(code_function_call.function()));
+
+    for(code_function_callt::argumentst::iterator
+          a_it=code_function_call.arguments().begin();
+        a_it!=code_function_call.arguments().end();
+        a_it++)
+      block.copy_to_operands(instrument_expr(*a_it));
+
+    block.copy_to_operands(code);
+    code=block;
+  }
+}
+
+/// Computes the instrumentation for `expr` in the form of
+/// either assertions or runtime exceptions.
+/// \par parameters: `expr`: the expression for which we compute
+/// strumentation \return: The instrumentation required for `expr`
+codet java_bytecode_instrumentt::instrument_expr(
+  const exprt &expr)
+{
+  if(expr.id()==ID_plus)
+  {
+    // this correspond to ?astore so we check that 0<=index<array.length
+    const plus_exprt &plus_expr=to_plus_expr(expr);
+    if(plus_expr.op0().id()==ID_member)
+    {
+      const member_exprt &member_expr=to_member_expr(plus_expr.op0());
+      if(member_expr.op0().id()==ID_dereference)
+      {
+        const dereference_exprt &dereference_expr=
+          to_dereference_expr(member_expr.op0());
+        codet bounds_check=
+          check_array_access(
+            dereference_expr,
+            plus_expr.op1(),
+            expr.source_location());
+        return bounds_check;
+      }
+    }
+  }
+  else if(expr.id()==ID_side_effect)
+  {
+    const side_effect_exprt &side_effect_expr=to_side_effect_expr(expr);
+    const irep_idt &statement=side_effect_expr.get_statement();
+    if(statement==ID_throw)
+      // this corresponds to athrow and so we check that
+      // we don't throw null
+      return check_null_dereference(
+        expr.op0(),
+        expr.source_location(),
+        true);
+    else if(statement==ID_java_new_array)
+      // this correpond to new array so we check that
+      // length is >=0
+      return check_array_length(
+        expr.op0(),
+        expr.source_location());
+  }
+  if(expr.id()==ID_member)
+  {
+    // this corresponds to either getfield or putfield
+    // so we must check null pointer dereference
```

Comment:
Note this also triggers (correctly) for ?astore, ?aload which synthesise a member_exprt, but will trigger twice (once for ->length, one for ->data). Might be simplest to have the frontend tag where null checks ought to be, in the style of the above suggested is-array-access tag?